src/ray/gcs/gcs_server/gcs_job_manager.cc

```diff
@@ -82,28 +86,37 @@ void GcsJobManager::HandleAddJob(rpc::AddJobRequest request,
   auto time = current_sys_time_ms();
   mutable_job_table_data.set_start_time(time);
   mutable_job_table_data.set_timestamp(time);
-  JobID job_id = JobID::FromBinary(mutable_job_table_data.job_id());
+  const JobID job_id = JobID::FromBinary(mutable_job_table_data.job_id());
   RAY_LOG(INFO) << "Adding job, job id = " << job_id
                 << ", driver pid = " << mutable_job_table_data.driver_pid();
 
-  auto on_done = [this, job_id, mutable_job_table_data, reply, send_reply_callback](
-                     const Status &status) {
+  auto on_done = [this,
+                  job_id,
+                  job_table_data_ref = std::cref(mutable_job_table_data),
```

Comment:
this keeps a reference to the local variable `mutable_job_table_data`, but the whole lambda `on_done` is put to an async `Put()` and to be invoked on a different stack, causing UAF. Please revert this. I agree we can rename it in the lambda like 

```
[...
job_table_data = mutable_job_table_data,
...](...){...}
```

Reply:
> this keeps a reference to the local variable mutable_job_table_data, but the whole lambda on_done is put to an async Put() and to be invoked on a different stack

Thanks for pointing that out; one thing I don't understand is we also capture `reply`, curious how we make sure it's also valid at callback invocation?

Reply:
Do you mind telling me what is `UAF`? I guess it's invalid memory access?

Reply:
And we also update `cached_job_configs_` in the callback without locking, curious how do we make sure no data race here?

Reply:
It's Use After Free - use of a dangling reference after it's destroyed, leading to undefined behavior.

Reply:
I think the callback will be called on the same thread by the ASIO io_context, so no lock is needed. If it's not the case we will see it as a CI tsan failure.

Reply:
since we already have tsan test in CI ... I hope any additional checkers be zero runtime cost to get merged.

Reply:
Don't want to argue, but the validity for TSAN completely depends on UT coverage :(

>  I hope any additional checkers be zero runtime cost to get merged.

It's super lightweight, just two atomic operations; or I could make it a DCHECK which does no-op at release.
Will get a PR later and let you review :)

Reply:
PR here: https://github.com/ray-project/ray/pull/47966